File: Application/src/main/java/com/shareyourproxy/api/rx/command/SaveGroupContactsCommand.java
package com.shareyourproxy.api.rx.command;

import android.app.Service;
import android.os.Parcel;

import com.shareyourproxy.api.domain.model.GroupToggle;
import com.shareyourproxy.api.domain.model.User;
import com.shareyourproxy.api.rx.RxGroupContactSync;
import com.shareyourproxy.api.rx.command.eventcallback.EventCallback;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Evan on 6/8/15.
 */
public class SaveGroupContactsCommand extends BaseCommand {
    public static final Creator<SaveGroupContactsCommand> CREATOR =
        new Creator<SaveGroupContactsCommand>() {
            @Override
            public SaveGroupContactsCommand createFromParcel(Parcel in) {
                return new SaveGroupContactsCommand(in);
            }

            @Override
            public SaveGroupContactsCommand[] newArray(int size) {
                return new SaveGroupContactsCommand[size];
            }
        };
    private final static java.lang.ClassLoader CL =
        SaveGroupContactsCommand.class.getClassLoader();

    public final ArrayList<GroupToggle> groups;
    public final User contact;
    public final User user;

    public SaveGroupContactsCommand( User user, ArrayList<GroupToggle> groups, User contact) {
        this.user = user;
        this.groups = groups;
        this.contact = contact;
    }

    private SaveGroupContactsCommand(Parcel in) {
        this((User) in.readValue(CL),
            (ArrayList<GroupToggle>) in.readValue(CL), (User) in.readValue(CL));
    }

    @Override
    public List<EventCallback> execute(Service service) {
        return RxGroupContactSync
            .updateGroupContacts(service, user, groups, contact);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeValue(user);
        dest.writeValue(groups);
        dest.writeValue(contact);
    }

}

File: Application/src/main/kotlin/com/shareyourproxy/api/rx/command/SaveGroupContactsCommand.kt
package com.shareyourproxy.api.rx.command

import android.content.Context
import android.os.Parcel
import android.os.Parcelable
import com.shareyourproxy.api.domain.model.GroupToggle
import com.shareyourproxy.api.domain.model.User
import com.shareyourproxy.api.rx.RxGroupContactSync.updateGroupContacts
import com.shareyourproxy.api.rx.command.eventcallback.EventCallback
import java.util.*

/**
 * Save contact to group.
 */
internal final class SaveGroupContactsCommand(val user: User, val groups: ArrayList<GroupToggle>, val contact: User) : BaseCommand() {
    @Suppress("UNCHECKED_CAST")
    private constructor(parcel: Parcel) : this(parcel.readValue(CL) as User, parcel.readValue(CL) as ArrayList<GroupToggle>, parcel.readValue(CL) as User)

    override fun execute(context: Context): EventCallback {
        return updateGroupContacts(context, user, groups, contact)
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeValue(user)
        dest.writeValue(groups)
        dest.writeValue(contact)
    }

    companion object {
        private val CL = SaveGroupContactsCommand::class.java.classLoader
        @JvmField val CREATOR: Parcelable.Creator<SaveGroupContactsCommand> = object : Parcelable.Creator<SaveGroupContactsCommand> {
            override fun createFromParcel(parcel: Parcel): SaveGroupContactsCommand {
                return SaveGroupContactsCommand(parcel)
            }

            override fun newArray(size: Int): Array<SaveGroupContactsCommand?> {
                return arrayOfNulls(size)
            }
        }
    }

}
